client/database.go

```diff
@@ -17,107 +17,212 @@ package client
-// within the Transact() method which abstracts away those operations.
-type Tx interface {
-	driver.Database
+// Database is the interface for interacting with a Tigris Database
+// Due to the limitations of Golang generics instantiations of the colletions
+// should be done using GetCollection[Model](ctx, db) top level function instead of
+// method of this interface.
+// Similarly to get access to collection APIs in a transaction
+// top level GetTxCollection(ctx, tx) function should be used
+// instead of method of Tx interface
+type Database struct {
+	name    string
+	schemas map[string]*schema.Schema
+	created map[string]bool
+	driver  driver.Driver
 }
 
-// Database is the interface for interacting with a specific database
-// in Tigris.
-type Database interface {
-	// Transact runs the provided TransactionFunc in a transaction. If the
-	// function returns an error then the transaction will be aborted,
-	// otherwise it will be committed.
-	Transact(ctx context.Context, fn TxFunc) (interface{}, error)
-
-	CreateCollections(ctx context.Context, dbName string, schemas []*schema.Schema) error
-}
-
-type database struct {
-	name   string
-	driver driver.Driver
-}
-
-func newDatabase(name string, driver driver.Driver) Database {
-	return &database{
-		name:   name,
-		driver: driver,
+func newDatabase(name string, driver driver.Driver) *Database {
+	return &Database{
+		name:    name,
+		driver:  driver,
+		schemas: make(map[string]*schema.Schema),
+		created: make(map[string]bool),
 	}
 }
 
-func (db *database) Transact(
-	ctx context.Context,
-	fn TxFunc,
-) (interface{}, error) {
-	tx, err := db.driver.BeginTx(ctx, db.name, nil)
-	if err != nil {
-		return nil, fmt.Errorf("error beginning transaction: %w", err)
-	}
-
-	res, err := fn(ctx, tx)
-	if err != nil {
-		if rollbackErr := tx.Rollback(ctx); rollbackErr != nil {
-			return nil, fmt.Errorf(
-				"error trying to rollback transaction: %v, original error: %w",
-				rollbackErr, err)
-		}
-		return nil, fmt.Errorf("error running transaction: %w", err)
-	}
-
-	return res, nil
-}
-
-func (db *database) CreateCollectionsFromModels(ctx context.Context, dbName string, model interface{}, models ...interface{}) error {
-	schemas, err := schema.FromCollectionModels(model, models)
+// CreateCollections creates collections in the Database using provided collection models
+func (db *Database) CreateCollections(ctx context.Context, model schema.Model, models ...schema.Model) error {
+	schemas, err := schema.FromCollectionModels(model, models...)
 	if err != nil {
 		return fmt.Errorf("error parsing model schema: %w", err)
 	}
 
-	return db.CreateCollections(ctx, dbName, schemas)
+	db.schemas = schemas
+
+	return db.createCollectionsFromSchemas(ctx, db.name, schemas)
 }
 
-func (db *database) CreateCollectionsFromDatabaseModel(ctx context.Context, dbModel interface{}) error {
+func (db *Database) createCollectionsFromDatabaseModel(ctx context.Context, dbModel schema.Model) error {
```

Comment:
Can you add an example of what the database model will look like? It will have all the collections?

Reply:
Yes.
type Db1{
   Coll1
   Coll2
   ...
}